File: 2685-count-the-number-of-complete-components/2685-count-the-number-of-complete-components.py
class Solution:
    def bfs(self, node: int, visited: List[int], graph: List[List[int]]) -> int:
        q = deque()
        q.append(node)

        visited[node] = 1

        nodeCount = edgeCount = 0
        while len(q) > 0:
            node = q.popleft()
            nodeCount += 1

            for adjNode in graph[node]:
                edgeCount += 1
                if visited[adjNode] == 1:
                    continue
                q.append(adjNode)
                visited[adjNode] = 1
        
        return int((edgeCount/2) == (nodeCount*(nodeCount-1)/2))

    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        #step 1 -> create the graph
        graph = []
        for i in range(n):
            graph.append([])
        
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        
        #step 2 -> do the bfs
        count = 0
        visited = [0] * n
        for node in range(n):
            if visited[node] == 0:
                count += self.bfs(node, visited, graph)
        
        return count

File: 2685-count-the-number-of-complete-components/2685-count-the-number-of-complete-components.cpp
/*
    1 - 2
    | \ |
    3.- 4
*/
class Solution {
public:
    int bfs(int node, vector<int>& visited, vector<vector<int>>& graph){
        queue<int> q;
        q.push(node);

        visited[node] = 1;

        int nodeCount = 0;
        int edgeCount = 0;
        while(!q.empty()){
            node = q.front();
            q.pop();

            nodeCount += 1;

            for(auto adj: graph[node]){
                edgeCount += 1;
                if(visited[adj] == 1){
                    continue;
                }

                q.push(adj);
                visited[adj] = 1;
            }
        }

        //check the full component here
        return ((edgeCount/2) == (nodeCount * (nodeCount-1)/2));
    }
    int countCompleteComponents(int n, vector<vector<int>>& edges) {
        //step #1: create the graph
        vector<vector<int>> graph(n);
        for(auto it: edges){
            graph[it[0]].push_back(it[1]);
            graph[it[1]].push_back(it[0]);
        }

        int count = 0;
        vector<int> visited(n, 0);
        //step #2: traverse the graph for each component alongside with their edge count
        //step #3: use the formula edgeCount == (n * (n-1)/ 2)
        for(int i = 0; i < n; i++){
            if(visited[i] == 0){
                count += bfs(i, visited, graph);
            }
        }

        return count;
    }
};